src/topology/uniform_space/basic.lean

```diff
@@ -1096,6 +1122,26 @@ lemma uniform_continuous_subtype_mk {p : α → Prop} [uniform_space α] [unifor
   uniform_continuous (λx, ⟨f x, h x⟩ : β → subtype p) :=
 uniform_continuous_comap' hf
 
+lemma uniform_continuous_on_iff_restrict [uniform_space α] [uniform_space β] (f : α → β) (s : set α) :
+uniform_continuous_on f s ↔ uniform_continuous (s.restrict f) :=
+begin
+  -- All the mathematical content is in `subtype_coe_map_comap` and functorial properties
+  -- of filter `map` and `comap`, but here we'll have to fight coercions in the two unnamed
+  -- intermediate facts
+  unfold uniform_continuous_on set.restrict uniform_continuous,
+  have :  (λ x : s × s, (f x.1, f x.2)) = (prod.map f f) ∘ (prod.map coe coe),
+  ext x ; refl,
+  rw [uniformity_comap rfl, this, tendsto, tendsto, ← filter.map_map, ← subtype_coe_map_comap],
+  convert iff.rfl using 3,
+  change map (prod.map coe coe) (filter.comap (prod.map coe coe) _) = _,
+  let φ : ↥s × ↥s → (s.prod s)  := (λ x : s × s, ⟨(x.1.1, x.2.1), mk_mem_prod x.1.2 x.2.2⟩),
+  have : (prod.map coe coe : s × s → α × α) = coe ∘ φ,
+  ext ; refl,
+  rw [this, ← filter.map_map, ← filter.comap_comap_comp, map_comap_of_surjective],
+  rintro ⟨⟨x, y⟩, ⟨x_in, y_in⟩⟩,
+  use [(⟨x, x_in⟩, ⟨y, y_in⟩), rfl],
+end
+
```

Comment:
```suggestion
lemma subtype_coe_map_comap_prod (s : set α) (f : filter (α × α)) :
  map (coe : s × s → α × α) (comap (coe : s × s → α × α) f) = f ⊓ 𝓟 (s.prod s) :=
let φ (x : s × s) : s.prod s := ⟨⟨x.1.1, x.2.1⟩, ⟨x.1.2, x.2.2⟩⟩ in
begin
  rw show (coe : s × s → α × α) = coe ∘ φ, by ext x; cases x; refl,
  rw [← filter.map_map, ← filter.comap_comap_comp],
  rw map_comap_of_surjective,
  exact subtype_coe_map_comap _ _,
  exact λ ⟨⟨a, b⟩, ⟨ha, hb⟩⟩, ⟨⟨⟨a, ha⟩, ⟨b, hb⟩⟩, rfl⟩
end

lemma uniform_continuous_on_iff_restrict [uniform_space α] [uniform_space β] (f : α → β) (s : set α) :
uniform_continuous_on f s ↔ uniform_continuous (s.restrict f) :=
begin
  -- All the mathematical content is in `subtype_coe_map_comap` and functorial properties
  -- of filter `map` and `comap`, but here we'll have to fight coercions in the two unnamed
  -- intermediate facts
  unfold uniform_continuous_on set.restrict uniform_continuous tendsto,
  rw [show (λ x : s × s, (f x.1, f x.2)) = prod.map f f ∘ coe, by ext x; cases x; refl],
  rw uniformity_comap rfl,
  rw [show prod.map subtype.val subtype.val = (coe : s × s → α × α), by ext x; cases x; refl],
  conv in (map _ (comap _ _)) { rw ← filter.map_map },
  rw subtype_coe_map_comap_prod, refl,
end
```

Wow, this is ugly.

Reply:
Oh yes it is. But your suggestion hides it better, so I'll use that. I can't simply accept the suggestion since things need to move to other files.